xarray/core/indexing.py

```diff
@@ -473,30 +480,64 @@ def _ensure_ndarray(self, value):
             value = utils.to_0d_array(value)
         return value
 
+    def _indexing_array_and_key(self, key):
+        if isinstance(key, OuterIndexer):
+            key = _outer_to_numpy_indexer(key, self.array.shape)
+
+        if isinstance(key, VectorizedIndexer):
+            array = nputils.NumpyVIndexAdapter(self.array)
+        else:
+            array = self.array
+
+        return array, to_tuple(key)
+
     def __getitem__(self, key):
-        key = self._convert_key(key)
-        return self._ensure_ndarray(self.array[key])
+        array, key = self._indexing_array_and_key(key)
+        return self._ensure_ndarray(array[key])
 
     def __setitem__(self, key, value):
-        key = self._convert_key(key)
-        self.array[key] = value
+        array, key = self._indexing_array_and_key(key)
+        array[key] = value
 
 
 class DaskIndexingAdapter(utils.NDArrayMixin):
-    """Wrap a dask array to support orthogonal indexing
+    """Wrap a dask array to support xarray-style indexing.
     """
     def __init__(self, array):
+        """ This adapter is usually called in Variable.__getitem__ with
+        array=Variable._broadcast_indexes
+        """
         self.array = array
 
     def __getitem__(self, key):
-        key = expanded_indexer(key, self.ndim)
-        if any(not isinstance(k, integer_types + (slice,)) for k in key):
-            value = self.array
-            for axis, subkey in reversed(list(enumerate(key))):
-                value = value[(slice(None),) * axis + (subkey,)]
+        def to_int_tuple(key):
+            # workaround for uint64 indexer (GH:1406)
```

Comment:
@jhamman 
I noticed dask array does not support uint indexer either.
I added a workaround here, but maybe it should be fixed in upstream.